drivers/serial/uart_bt.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2024 Croxel, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/bluetooth/services/nus.h>
+
+#define DT_DRV_COMPAT zephyr_nus_uart
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uart_nus, CONFIG_UART_LOG_LEVEL);
+
+struct uart_bt_data {
+	struct {
+		struct bt_nus_inst *inst;
+		struct bt_nus_cb cb;
+		atomic_t enabled;
+	} bt;
+	struct {
+		struct ring_buf *rx_ringbuf;
+		struct ring_buf *tx_ringbuf;
+		struct k_work cb_work;
+		struct k_work_delayable tx_work;
+		bool rx_irq_ena;
+		bool tx_irq_ena;
+		struct {
+			const struct device *dev;
+			uart_irq_callback_user_data_t cb;
+			void *cb_data;
+		} callback;
+	} uart;
+};
+
+static void bt_notif_enabled(bool enabled, void *ctx)
+{
+	__ASSERT_NO_MSG(ctx);
+
+	const struct device *dev = (const struct device *)ctx;
+	struct uart_bt_data *dev_data = (struct uart_bt_data *)dev->data;
+
+	(void)atomic_set(&dev_data->bt.enabled, enabled ? 1 : 0);
+
+	LOG_DBG("%s() - %s", __func__, enabled ? "enabled" : "disabled");
+
+	if (!ring_buf_is_empty(dev_data->uart.tx_ringbuf)) {
+		k_work_reschedule(&dev_data->uart.tx_work, K_NO_WAIT);
+	}
+}
+
+static void bt_received(struct bt_conn *conn, const void *data, uint16_t len, void *ctx)
+{
+	__ASSERT_NO_MSG(conn);
+	__ASSERT_NO_MSG(ctx);
+	__ASSERT_NO_MSG(data);
+	__ASSERT_NO_MSG(len > 0);
+
+	const struct device *dev = (const struct device *)ctx;
+	struct uart_bt_data *dev_data = (struct uart_bt_data *)dev->data;
+	struct ring_buf *ringbuf = dev_data->uart.rx_ringbuf;
+	uint32_t put_len;
+
+	LOG_DBG("%s() - len: %d, rx_ringbuf space %d", __func__, len, ring_buf_space_get(ringbuf));
+	LOG_HEXDUMP_DBG(data, len, "data");
+
+	put_len = ring_buf_put(ringbuf, (const uint8_t *)data, len);
+	if (put_len < len) {
+		LOG_ERR("RX Ring buffer full. received: %d, added to queue: %d", len, put_len);
+	}
+
+	k_work_submit(&dev_data->uart.cb_work);
+}
+
+static void cb_work_handler(struct k_work *work)
+{
+	struct uart_bt_data *dev_data = CONTAINER_OF(work, struct uart_bt_data, uart.cb_work);
+
+	if (dev_data->uart.callback.cb) {
+		dev_data->uart.callback.cb(
+			dev_data->uart.callback.dev,
+			dev_data->uart.callback.cb_data);
+	}
+}
+
+static void tx_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct uart_bt_data *dev_data = CONTAINER_OF(dwork, struct uart_bt_data, uart.tx_work);
+	uint8_t *data = NULL;
+	size_t len;
+	int err;
+
+	__ASSERT_NO_MSG(dev_data);
+
+	do {
+		/** Using Minimum MTU at this point to guarantee all connected
+		 * peers will receive the data, without keeping track of MTU
+		 * size per-connection. This has the trade-off of limiting
+		 * throughput but allows multi-connection support.
+		 */
+		len = ring_buf_get_claim(dev_data->uart.tx_ringbuf, &data, 23);
```

Comment:
We must subtract the ATT_HANDLE_VALUE_NTF overhead (3 bytes) from the minimum ATT_MTU (23 bytes) to get the useful MTU.

```suggestion
		len = ring_buf_get_claim(dev_data->uart.tx_ringbuf, &data, 20);
```

Reply:
Good catch - Thanks for pointing it out. I can address it on a follow-up PR if that's ok!

Reply:
Addressed by #70668. Resolving now.